lib/ddtrace/contrib/sinatra/patcher.rb

```diff
@@ -1,9 +1,32 @@
 # typed: true
 require 'ddtrace/contrib/patcher'
+require 'ddtrace/contrib/sinatra/framework'
+require 'ddtrace/contrib/rack/middlewares'
 
 module Datadog
   module Contrib
     module Sinatra
+      # Set tracer configuration at a late enough time
+      module TracerSetupPatch
+        def setup_middleware(*args, &block)
+          super.tap do
+            Datadog::Contrib::Sinatra::Framework.setup
```

Comment:
I don't know if this is the best place to call this, the Rails integration does it on `after_initialize` but I don't think there's any equivalent for Sinatra, short of patching another method. Any ideas/opinions regarding the lifecycle of the configuration?

Reply:
I also saw a bunch of "only once" stuff over there for Rails. Another concern I have is whether there exists the possibility of having a Rails app that mounts Sinatra on a route, and maybe the Sinatra integration gets setup first, then the `rack.request` spans would have Sinatra as a service name instead of Rails?

I lack insight of the config/initialization lifecycle to pursue this question further. Any ideas?

Reply:
> I also saw a bunch of "only once" stuff over there for Rails. Another concern I have is whether there exists the possibility of having a Rails app that mounts Sinatra on a route, and maybe the Sinatra integration gets setup first, then the rack.request spans would have Sinatra as a service name instead of Rails?

I don't think we can make a correct authoritative decision on service name in the case where two frameworks (Rails & Sinatra) trying to back fill Rack. In such case, the user will have to specify Rack's configuration explicitly. You can leave the fallback implemented in this PR as is.

> I lack insight of the config/initialization lifecycle to pursue this question further. Any ideas?

I think that we should have a `only once` guard for              `Datadog::Contrib::Sinatra::Framework.setup`. Currently it is safe to run many times (because it only conditionally configures Rack), but that might not be the case in the future.
Also, we only need to configure the fallback Rack integration once per application lifecycle.

Reply:
Added OnlyOnce